src/pjpersist/tests/test_datamanager.py

```diff
@@ -1612,10 +1613,74 @@ def test_dirty_dm_tpc_prepare(self):
         self.assertTrue('tpc_prepare' in str(prep_mock_p.call_args_list))
         self.assertTrue('tpc_commit' in str(prep_mock_p.call_args_list))
 
+    def test_release_on_destroy(self):
+        # When DM is destroyed without committing or aborting the transaction,
+        # connection should be released.
+        # GIVEN
+        conn = testing.getConnection(testing.DBNAME)
+        pool = testing.DummyConnectionPool(conn)
+        dm = datamanager.PJDataManager(pool)
+        dm.begin()
+        self.assertTrue(pool.isTaken())
+
+        # WHEN
+
+        # Create new transaction to release the reference to our data manager
+        # and allow it to be destroyed
+        transaction.begin()
+
+        # THEN
+        self.assertFalse(pool.isTaken())
+
+    def test_release_on_commit(self):
+        # GIVEN
+        transaction.abort()
+        pool = testing.DummyConnectionPool(self.conn)
+        dm = datamanager.PJDataManager(pool)
+        dm.begin()
+        self.assertTrue(pool.isTaken())
+
+        # WHEN
+
+        # Create new transaction to release the reference to our data manager
+        # and allow it to be destroyed
+        transaction.commit()
+
+        # THEN
+        self.assertFalse(pool.isTaken())
+
+    def test_release_on_abort(self):
+        # GIVEN
+        transaction.abort()
+        pool = testing.DummyConnectionPool(self.conn)
+        dm = datamanager.PJDataManager(pool)
+        dm.begin()
+        self.assertTrue(pool.isTaken())
+
+        # WHEN
+
+        # Create new transaction to release the reference to our data manager
+        # and allow it to be destroyed
+        transaction.abort()
+
+        # THEN
+        self.assertFalse(pool.isTaken())
+
+    def test_readonly_fail_on_write(self):
+        transaction.abort()
+        pool = testing.DummyConnectionPool(self.conn)
+        dm = datamanager.PJDataManager(pool)
+        dm.setTransactionOptions(readonly=True)
+
+        # WHEN, THEN
+        with self.assertRaises(interfaces.ReadOnlyDataManagerError):
+            with dm.getCursor() as cur:
+                cur.execute("INSERT INTO mytab VALUES (1, '10')")
```

Comment:
What's up with the connection here? Should it get returned?

Reply:
It doesn't matter for the test - it uses `DummyConnectionPool`.

Reply:
I think it matters in the sense of I'd expect that the manager returns the connection on an error

Reply:
This is a test about handling inserts in readonly transaction. It doesn't test how connections return. Connection gets returned when transaction is aborted or committed - testing that is out of scope of this particular test.

Reply:
So you mean execute fails, transaction gets aborted, abort returns the connection

Reply:
Yes, transaction is aborted by test infrastructure in any case (success or failure)